sofar/sofar.py

```diff
@@ -0,0 +1,2125 @@
+        required by the SOFA definition. Note that the shape is the shape of
+        the array that holds the strings. NETCDF stores all string variables in
+        arrays.
+        """
+
+        if isinstance(value, str):
+            S = len(value)
+            shape = (1, 1)
+        elif isinstance(value, list):
+            S = len(max(value, key=len))
+            shape = np.array(value).shape
+        elif isinstance(value, np.ndarray):
+            S = max(np.vectorize(len)(value))
+            shape = value.shape
+        else:
+            raise TypeError((f"{key} must be a string, numpy string array, "
+                             "or list of strings"))
+
+        return S, shape
+
+    @staticmethod
+    def _mandatory(flags):
+        """
+        Check if a field is mandatory
+
+        Parameters
+        ----------
+        flags : None, str
+            The flags from convention[key]["flags"]
+
+        Returns
+        -------
+        is_mandatory : bool
+        """
+        # skip optional fields if requested
+        if flags is None:
+            is_mandatory = False
+        elif "m" not in flags:
+            is_mandatory = False
+        else:
+            is_mandatory = True
+
+        return is_mandatory
+
+    @staticmethod
+    def _read_only(flags):
+        """
+        Check if a field is read only
+
+        Parameters
+        ----------
+        flags : None, str
+            The flags from convention[key]["flags"]
+
+        Returns
+        -------
+        is_read_only : bool
+        """
+        # skip optional fields if requested
+        if flags is None:
+            is_read_only = False
+        elif "r" not in flags:
+            is_read_only = False
+        else:
+            is_read_only = True
+
+        return is_read_only
+
+
+def update_conventions(conventions_path=None):
+    """
+    Update SOFA conventions.
+
+    A SOFA convention defines the kind of data and the data format that is
+    stored in a SOFA file. Updating the conventions is done in two steps:
+
+    1.
+        Download official SOFA conventions as csv files from
+        https://github.com/sofacoustics/API_MO/tree/master/API_MO/conventions.
+    2.
+        Convert csv files to json files for easier handling
+
+    The csv and json files are stored at sofar/conventions. Sofar works only on
+    the json files. To get a list of all currently available SOFA conventions
+    and their paths see :py:func:`~sofar.list_conventions`.
+
+    .. note::
+        If the official convention contain errors, calling this function might
+        break sofar. Be sure that you want to do this.
+
+    Parameters
+    ----------
+    conventions_path : str, optional
+        Path to the folder where the conventions are saved. The default is
+        ``None``, which saves the conventions inside the sofar package.
+        Conventions saved under a different path can not be used by sofar.
+    """
+
```

Comment:
This is related to a previous comment on the generation of the JSON files:
I'm pretty unsure on how to deal with this.

1. I like the idea that the convention JSONs are generated from the csv files, which does make storing the JSONs inside the repository redundant however.
2. Since the csv files are downloaded using this file, storing them inside the repository is again redundant.
3. Issues may occur if conventions are updated by the user, which break functionality of older package versions, which do not officially support these versions.
4. Depending on where the package is installed to, updating the conventions will require root/Admin access, which I would avoid due to security considerations/usability.

Solutions could be:
1. Storing the conventions in a user setting folder inside the users home folder (can be found using the PATH variable usually). This avoids file system issues as well as multi-user installs on shared machines (even though thats discouraged in favor of venvs anyways, it should still not lead to issues to begin with)

Reply:
As discussed - keep both but in separate folders.